os/fs/smartfs/smartfs_utils.c

```diff
@@ -1716,139 +1962,255 @@ int smartfs_countdirentries(struct smartfs_mountpt_s *fs, struct smartfs_entry_s
+			goto errout_with_filepos;
 		}
+		nextsector = sector;
+	}
 
-		/* Get the next chained sector */
+errout_with_filepos:
+	sf->filepos = savefilepos;
+	return ret;
+}
 
-		sector = SMARTFS_NEXTSECTOR(header);
+/****************************************************************************
+ * Name: smartfs_extendfile
+ *
+ * Description:
+ *   Zero-extend the length of a regular file to 'length'.
+ *
+ ****************************************************************************/
 
-		/* If this is the 1st sector of the file, then just overwrite
-		 * the sector data with the erased state value.  The underlying
-		 * SMART block driver will detect this and release the old
-		 * sector and create a new one with the new (blank) data.
-		 */
+int smartfs_extendfile(FAR struct smartfs_mountpt_s *fs, FAR struct smartfs_ofile_s *sf, off_t length)
+{
+	int ret;
+	int result;
+	uint16_t nextsector;
+	struct smart_read_write_s readwrite;
+	struct smartfs_chain_header_s *chainheader;
+	uint16_t savefilepos;
+	uint16_t data_len;
+	uint16_t remaining;
+	uint16_t offset;
+	int next_chain;
+	uint16_t save_first;
+	bool new_chain = false;
 
-		if (nextsector == entry->firstsector) {
-#ifdef CONFIG_SMARTFS_USE_SECTOR_BUFFER
+	/* Sector data length does not include size of the chainheader */
+	data_len = fs->fs_llformat.availbytes - sizeof(struct smartfs_chain_header_s);
 
-			/* When we have a sector buffer in use, simply skip the first sector */
+	remaining = length - sf->entry.datalen;
+	offset = sizeof(struct smartfs_chain_header_s);
+	savefilepos = sf->filepos;
+	save_first = 0xFFFF;
+	result = OK;
 
-			nextsector = sector;
-			continue;
+	/* Seek to the current end of the file, add '0's after this to extend the length of the file */
+	smartfs_seek_internal(fs, sf, 0, SEEK_END);
 
-#else
+	if (remaining <= (fs->fs_llformat.availbytes - sf->curroffset)) {
+		/* Case 1:- the length extension can be done in the current last sector itself */
+		goto extend_last_sector;
+	} else {
+		/* Case 2:- need to add new sectors to file to extend the length */
+		new_chain = true;
+		/* We will add extra bytes to current last sector later. First we will allocate and write the extra sectors required. */
+		remaining -= (fs->fs_llformat.availbytes - sf->curroffset);
+		ret = FS_IOCTL(fs, BIOC_ALLOCSECT, 0xFFFF);
+		if (ret < 0) {
+			fdbg("Unable to allocate new sector for extending file, error = %d\n", ret);
+			return ret;
+		}
+		save_first = (uint16_t)ret;
 
-			/* Fill our buffer with erased data */
+		while (ret != SMARTFS_ERASEDSTATE_16BIT) {
+			next_chain = SMARTFS_ERASEDSTATE_16BIT;
+			chainheader = (struct smartfs_chain_header_s *)fs->fs_rwbuffer;
+			chainheader->type = SMARTFS_DIRENT_TYPE_FILE;
 
-			memset(fs->fs_rwbuffer, CONFIG_SMARTFS_ERASEDSTATE, fs->fs_llformat.availbytes);
-			header->type = SMARTFS_SECTOR_TYPE_FILE;
+			if (remaining >= data_len) {
+				memset(&fs->fs_rwbuffer[offset], 0, data_len);
```

Comment:
According to POSIX, '\0' is correct. As I know 0 and '\0' are different(most of platform it is same), so we'd better use it properly.

Reply:
Yes, will check